cmd/frontend/graphqlbackend/search_results.go

```diff
@@ -520,95 +520,98 @@ func (r *searchResolver) evaluateLeaf(ctx context.Context) (_ *SearchResultsReso
+
+	if fileMatch, ok := r.ToFileMatch(); ok {
+		if prev, seen := d.seenFileMatches[fileMatch.uri]; seen {
+			prev.appendMatches(fileMatch)
+			return
 		}
+		d.seenFileMatches[fileMatch.uri] = fileMatch
+		return
+	}
 
-		if leftRepoMatch, ok := leftMatch.ToRepository(); ok {
-			rightRepoMatch := rightRepoMatches[string(leftRepoMatch.URL())]
-			if rightRepoMatch == nil {
-				// no overlap with existing matches.
-				merged = append(merged, leftMatch)
-				count++
-			}
-			continue
+	if repoMatch, ok := r.ToRepository(); ok {
+		if _, seen := d.seenRepoMatches[repoMatch.URL()]; seen {
+			return
 		}
+		d.seenRepoMatches[repoMatch.URL()] = repoMatch
+		return
+	}
 
-		if leftCommitMatch, ok := leftMatch.ToCommitSearchResult(); ok {
-			if leftCommitMatch.DiffPreview() != nil {
-				rightDiffMatch := rightDiffMatches[leftCommitMatch.URL()]
-				if rightDiffMatch == nil {
-					merged = append(merged, leftCommitMatch)
-					count++
-				}
-			} else {
-				rightCommitMatch := rightCommitMatches[leftCommitMatch.URL()]
-				if rightCommitMatch == nil {
-					merged = append(merged, leftCommitMatch)
-					count++
-				}
+	if commitMatch, ok := r.ToCommitSearchResult(); ok {
+		if commitMatch.DiffPreview() != nil {
+			if _, seen := d.seenDiffMatches[commitMatch.URL()]; seen {
+				return
 			}
-			continue
+			d.seenDiffMatches[commitMatch.URL()] = commitMatch
+			return
+		} else {
+			if _, seen := d.seenCommitMatches[commitMatch.URL()]; seen {
+				return
+			}
+			d.seenCommitMatches[commitMatch.URL()] = commitMatch
+			return
 		}
-		merged = append(merged, leftMatch)
 	}
 
-	for _, v := range rightFileMatches {
-		merged = append(merged, v)
+	return
+}
+
+// Deduped returns a slice of SearchResultResolvers, deduplicated from
+// the SearchResultResolvers that were added with Add
+func (d *searchResultDeduper) Deduped() []SearchResultResolver {
+	total := len(d.seenFileMatches) + len(d.seenRepoMatches) + len(d.seenCommitMatches) + len(d.seenDiffMatches)
+	r := make([]SearchResultResolver, 0, total)
+	for _, v := range d.seenFileMatches {
+		r = append(r, v)
 	}
-	for _, v := range rightRepoMatches {
-		merged = append(merged, v)
+	for _, v := range d.seenRepoMatches {
+		r = append(r, v)
 	}
-	for _, v := range rightCommitMatches {
-		merged = append(merged, v)
+	for _, v := range d.seenCommitMatches {
+		r = append(r, v)
 	}
-	for _, v := range rightDiffMatches {
-		merged = append(merged, v)
+	for _, v := range d.seenDiffMatches {
+		r = append(r, v)
 	}
+	return r
+}
 
-	left.SearchResults = merged
+// unionMerge performs a merge of file match results, merging line matches when
+// they occur in the same file.
+func unionMerge(left, right *SearchResultsResolver) *SearchResultsResolver {
+	var dedup searchResultDeduper
+
+	// Register all the left results in the dedupper
+	for _, leftResult := range left.SearchResults {
+		dedup.Add(leftResult)
+	}
+
+	for _, rightResult := range right.SearchResults {
+		dedup.Add(rightResult)
+	}
```

Comment:
```suggestion
	// Add results to maps for deduping.
	for _, leftResult := range left.SearchResults {
		dedup.Add(leftResult)
	}
	for _, rightResult := range right.SearchResults {
		dedup.Add(rightResult)
	}
```

Nit: comment seems unbalanced talking about left and not right :-)